File: 0862/solution.py
from collections import deque


class Solution:
    def shortestSubarray(self, nums: list[int], k: int) -> int:
        stack = deque()
        stack.append([-1, 0])
        res = float("inf")
        total = 0
        for i, v in enumerate(nums):
            if v >= k:
                return 1
            total += v
            while stack and total <= stack[-1][-1]:
                stack.pop()
            stack.append([i, total])
            while len(stack) >= 2 and stack[-1][-1] - stack[0][-1] >= k:
                res = min(res, stack[-1][0] - stack[0][0])
                stack.popleft()
        return res if res != float("inf") else -1

File: 0862/solution.java
import java.util.ArrayDeque;

class Solution {
    public int shortestSubarray(int[] nums, int k) {
        ArrayDeque<long[]> window = new ArrayDeque<>();
        window.add(new long[] { -1, 0 });
        long res = Integer.MAX_VALUE;
        long total = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] >= k) {
                return 1;
            }
            total += nums[i];
            while (!window.isEmpty() && window.getLast()[1] >= total) {
                window.removeLast();
            }
            window.add(new long[] { i, total });
            while (window.size() >= 2 && window.getLast()[1] - window.getFirst()[1] >= k) {
                res = Math.min(res, window.getLast()[0] - window.getFirst()[0]);
                window.removeFirst();
            }
        }
        return res == Integer.MAX_VALUE ? -1 : (int) res;
    }
}

// time O(n)
// space O(n)